director/director.go

```diff
@@ -384,14 +388,62 @@ func redirectToOrigin(ginCtx *gin.Context) {
 		ginCtx.String(http.StatusNotFound, "There are currently no origins exporting the provided namespace prefix")
 		return
 	}
+
+	availableOriginAds := []server_structs.ServerAd{}
+	// Skip stat query for PUT (upload), PROPFIND (listing) or skipStat query flag is on
+	if ginCtx.Request.Method == "PUT" || ginCtx.Request.Method == "PROPFIND" || skipStat {
+		availableOriginAds = originAds
+	} else {
+		// Query Origins and check if the object exists on the server
+		q := NewObjectStat()
+		qr := q.Query(context.Background(), reqPath, config.OriginType, 1, 3,
+			withOriginAds(originAds), WithToken(authzBearerEscaped))
+		log.Debugf("Stat result for %s: %s", reqPath, qr.String())
+
+		if qr.Status == querySuccessful {
+			for _, obj := range qr.Objects {
+				serverUrl := obj.URL.Host
+				for _, oAd := range originAds {
+					if oAd.URL.Host == serverUrl {
+						availableOriginAds = append(availableOriginAds, oAd)
+					}
+				}
+			}
+		} else if qr.Status == queryFailed {
+			if qr.ErrorType != queryInsufficientResErr {
+				ginCtx.String(http.StatusInternalServerError, fmt.Sprintf("Failed to query origins with error %s: %s", string(qr.ErrorType), qr.Msg))
+				return
+			}
+			// Insufficient response
+			if len(qr.DeniedServers) == 0 {
+				// No denied server, the object was not found on any origins
+				ginCtx.String(http.StatusNotFound, "There are currently no origins hosting the object")
+				return
+			}
+			// For denied servers, append them to availableOriginAds
+			for _, ds := range qr.DeniedServers {
+				for _, oAd := range originAds {
+					if oAd.AuthURL.String() == ds {
```

Comment:
Yes. I need to clarify in the comment section. We use `oAd.URL.Host` in the previous block because that's the object url, which is the data url, which is the key of originAds. In this block, the `DeniedServers` is a list of server AuthUrls, so we use `oAd.AuthURL` to match